// react-jinke-music-player/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = require("react-dom");

var _classnames = _interopRequireDefault(require("classnames"));

var _isMobile = _interopRequireDefault(require("is-mobile"));

var _Slider = _interopRequireDefault(require("rc-slider/lib/Slider"));

var _rcSwitch = _interopRequireDefault(require("rc-switch"));

var _downloadjs = _interopRequireDefault(require("downloadjs"));

var _reactDraggable = _interopRequireDefault(require("react-draggable"));

var _utils = require("./utils");

var _AudioListsPanel = _interopRequireDefault(require("./components/AudioListsPanel"));

var _PlayerMobile = _interopRequireDefault(require("./components/PlayerMobile"));

var _lyric = _interopRequireDefault(require("./lyric"));

var _Icon = require("./components/Icon");

var _CircleProcessBar = _interopRequireDefault(require("./components/CircleProcessBar"));

var _Load = _interopRequireDefault(require("./components/Load"));

var _PlayModel = _interopRequireDefault(require("./components/PlayModel"));

var _slider = require("./config/slider");

var _keycode = require("./config/keycode");

var _propTypes = _interopRequireDefault(require("./config/propTypes"));

var _locale = _interopRequireDefault(require("./config/locale"));

var _networkState = _interopRequireDefault(require("./config/networkState"));

var _animate = require("./config/animate");

var _playMode2 = _interopRequireDefault(require("./config/playMode"));

var _locale2 = _interopRequireDefault(require("./locale"));

require("rc-slider/assets/index.css");

require("rc-switch/assets/index.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var IS_MOBILE = (0, _isMobile["default"])();

var ReactJkMusicPlayer = /*#__PURE__*/function (_PureComponent) {
  _inherits(ReactJkMusicPlayer, _PureComponent);

  var _super = _createSuper(ReactJkMusicPlayer);

  _createClass(ReactJkMusicPlayer, [{
    key: "locale",
    //初始播放id
    get: function get() {
      var locale = this.props.locale;

      if (typeof locale === "string") {
        return _locale2["default"][this.props.locale];
      }

      return locale ? _objectSpread({}, _locale2["default"][_locale["default"].en_US], {}, locale) : {};
    }
  }]);

  function ReactJkMusicPlayer(props) {
    var _this;

    _classCallCheck(this, ReactJkMusicPlayer);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "initPlayId", "");

    _defineProperty(_assertThisInitialized(_this), "state", {
      audioLists: [],
      playId: _this.initPlayId,
      //播放id
      name: "",
      //当前歌曲名
      cover: "",
      //当前歌曲封面
      singer: "",
      //当前歌手
      musicSrc: "",
      //当前歌曲链
      lyric: "",
      // 当前歌词
      currentLyric: "",
      isMobile: IS_MOBILE,
      toggle: false,
      pause: true,
      playing: false,
      duration: 0,
      currentTime: 0,
      isLoop: false,
      isMute: false,
      soundValue: 100,
      moveX: 0,
      moveY: 0,
      isMove: false,
      loading: false,
      audioListsPanelVisible: false,
      playModelNameVisible: false,
      theme: _this.darkThemeName,
      extendsContent: null,
      //自定义扩展功能按钮
      playMode: "",
      //当前播放模式
      currentAudioVolume: 0,
      //当前音量  静音后恢复到之前记录的音量
      initAnimate: false,
      isInitAutoplay: false,
      isInitRemember: false,
      loadProgress: 0,
      removeId: -1,
      isNeedMobileHack: IS_MOBILE,
      audioLyricVisible: false,
      isAudioListsChange: false,
      notAutoPlayUntilPlayClicked: false
    });

    _defineProperty(_assertThisInitialized(_this), "getAudioTitle", function () {
      // 暂时兼容
      var _ref = _this.locale || {},
          audioTitle = _ref.audioTitle;

      var _this$state = _this.state,
          name = _this$state.name,
          singer = _this$state.singer;

      if (typeof audioTitle === "function" && _this.audio) {
        return audioTitle(_this.getBaseAudioInfo());
      }

      return audioTitle || "".concat(name, " ").concat(singer ? "- ".concat(singer) : "");
    });

    _defineProperty(_assertThisInitialized(_this), "toggleAudioLyric", function () {
      _this.setState({
        audioLyricVisible: !_this.state.audioLyricVisible
      });
    });

    _defineProperty(_assertThisInitialized(_this), "togglePlayMode", function () {
      var index = _this._PLAY_MODE_.findIndex(function (mode) {
        return mode === _this.state.playMode;
      });

      var playMode = index === _this._PLAY_MODE_LENGTH_ - 1 ? _this._PLAY_MODE_[0] : _this._PLAY_MODE_[++index];

      _this.setState({
        playMode: playMode,
        playModelNameVisible: true,
        playModeTipVisible: true
      });

      _this.props.onPlayModeChange && _this.props.onPlayModeChange(playMode);
      clearTimeout(_this.playModelTimer);
      _this.playModelTimer = setTimeout(function () {
        _this.setState({
          playModelNameVisible: false,
          playModeTipVisible: false
        });
      }, _this.props.playModeShowTime);
    });

    _defineProperty(_assertThisInitialized(_this), "renderPlayModeIcon", function (playMode) {
      var IconNode = "";
      var animateName = "react-jinke-music-player-mode-icon";

      switch (playMode) {
        case _playMode2["default"].order:
          IconNode = /*#__PURE__*/_react["default"].createElement(_Icon.OrderPlayIcon, {
            className: animateName
          });
          break;

        case _playMode2["default"].orderLoop:
          IconNode = /*#__PURE__*/_react["default"].createElement(_Icon.RepeatIcon, {
            className: animateName
          });
          break;

        case _playMode2["default"].singleLoop:
          IconNode = /*#__PURE__*/_react["default"].createElement(_Icon.LoopIcon, {
            className: animateName
          });
          break;

        case _playMode2["default"].shufflePlay:
          IconNode = /*#__PURE__*/_react["default"].createElement(_Icon.ShufflePlayIcon, {
            className: animateName
          });
          break;

        default:
          IconNode = /*#__PURE__*/_react["default"].createElement(_Icon.OrderPlayIcon, {
            className: animateName
          });
      }

      return IconNode;
    });

    _defineProperty(_assertThisInitialized(_this), "audioListsPlay", function (playId) {
      var ignore = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var state = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _this.state;
      var currentPlayId = state.playId,
          pause = state.pause,
          playing = state.playing,
          audioLists = state.audioLists;

      if (Array.isArray(audioLists) && audioLists.length === 0) {
        /*eslint-disable-next-line no-console*/
        return console.warn("Your playlist has no songs. and cannot play !");
      } //如果点击当前项 就暂停 或者播放


      if (playId === currentPlayId && !ignore) {
        _this.setState({
          pause: !pause,
          playing: !playing
        });

        return pause ? _this.audio.play() : _this._pauseAudio();
      }

      var _audioLists$find = audioLists.find(function (audio) {
        return audio.id === playId;
      }),
          name = _audioLists$find.name,
          cover = _audioLists$find.cover,
          musicSrc = _audioLists$find.musicSrc,
          singer = _audioLists$find.singer,
          _audioLists$find$lyri = _audioLists$find.lyric,
          lyric = _audioLists$find$lyri === void 0 ? "" : _audioLists$find$lyri;

      var loadAudio = function loadAudio(musicSrc) {
        _this.setState({
          name: name,
          cover: cover,
          musicSrc: musicSrc,
          singer: singer,
          playId: playId,
          lyric: lyric,
          currentTime: 0,
          duration: 0,
          playing: false,
          loading: true,
          loadProgress: 0
        }, function () {
          _this.initLyricParser();

          _this.audio.load();

          _this.updateMediaSessionMetadata();
        });

        _this.props.onAudioPlay && _this.props.onAudioPlay(_this.getBaseAudioInfo());
        _this.props.onAudioPlayTrackChange && _this.props.onAudioPlayTrackChange(playId, audioLists, _this.getBaseAudioInfo());
      };

      switch (_typeof(musicSrc)) {
        case "function":
          musicSrc().then(function (originMusicSrc) {
            loadAudio(originMusicSrc);
          }, _this.onAudioLoadError);
          break;

        default:
          loadAudio(musicSrc);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "resetAudioStatus", function () {
      _this.audio.pause();

      _this.initPlayInfo([]);

      _this.setState({
        currentTime: 0,
        duration: 0,
        loading: false,
        playing: false,
        pause: true,
        currentLyric: "",
        playId: _this.initPlayId
      });
    });

    _defineProperty(_assertThisInitialized(_this), "deleteAudioLists", function (audioId) {
      return function (e) {
        e.stopPropagation(); //如果不 传 id  删除全部

        var _this$state2 = _this.state,
            audioLists = _this$state2.audioLists,
            playId = _this$state2.playId;

        if (audioLists.length < 1) {
          return;
        }

        _this.lyric && _this.lyric.stop();

        if (!audioId) {
          _this.props.onAudioListsChange && _this.props.onAudioListsChange("", [], {});
          return _this.resetAudioStatus();
        }

        var newAudioLists = _toConsumableArray(audioLists).filter(function (audio) {
          return audio.id !== audioId;
        }); //触发删除动画,等动画结束 删除列表


        _this.setState({
          removeId: audioId
        });

        setTimeout(function () {
          _this.setState({
            audioLists: newAudioLists,
            removeId: -1
          }, function () {
            if (!newAudioLists.length) {
              return _this.resetAudioStatus();
            } // 如果删除的是当前正在播放的 顺延下一首播放


            if (audioId === playId) {
              _this.handlePlay(_playMode2["default"].orderLoop);
            }
          });
        }, _animate.AUDIO_LIST_REMOVE_ANIMATE_TIME);
        _this.props.onAudioListsChange && _this.props.onAudioListsChange(playId, newAudioLists, _this.getBaseAudioInfo());
      };
    });

    _defineProperty(_assertThisInitialized(_this), "openAudioListsPanel", function () {
      _this.setState(function (_ref2) {
        var audioListsPanelVisible = _ref2.audioListsPanelVisible;
        return {
          initAnimate: true,
          audioListsPanelVisible: !audioListsPanelVisible
        };
      });

      _this.props.onAudioListsPanelChange && _this.props.onAudioListsPanelChange(!_this.state.audioListsPanelVisible);
    });

    _defineProperty(_assertThisInitialized(_this), "closeAudioListsPanel", function (e) {
      e.stopPropagation();

      _this.setState({
        audioListsPanelVisible: false
      });

      _this.props.onAudioListsPanelChange && _this.props.onAudioListsPanelChange(false);
    });

    _defineProperty(_assertThisInitialized(_this), "themeChange", function (isLight) {
      var theme = isLight ? _this.lightThemeName : _this.darkThemeName;

      _this.setState({
        theme: theme
      });

      _this.props.onThemeChange && _this.props.onThemeChange(theme);
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioDownload", function () {
      var musicSrc = _this.state.musicSrc;

      if (_this.state.musicSrc) {
        var customDownloader = _this.props.customDownloader;

        var baseAudioInfo = _this.getBaseAudioInfo();

        var onBeforeAudioDownload = _this.props.onBeforeAudioDownload(baseAudioInfo);

        var transformedDownloadAudioInfo = {};

        if (onBeforeAudioDownload && onBeforeAudioDownload.then) {
          onBeforeAudioDownload.then(function (info) {
            var src = info.src,
                filename = info.filename,
                mimeType = info.mimeType;
            transformedDownloadAudioInfo = info;

            if (customDownloader) {
              customDownloader(info);
            } else {
              (0, _downloadjs["default"])(src, filename, mimeType);
            }
          });
        } else {
          customDownloader ? customDownloader({
            src: musicSrc
          }) : (0, _downloadjs["default"])(musicSrc);
        }

        _this.props.onAudioDownload && _this.props.onAudioDownload(baseAudioInfo, transformedDownloadAudioInfo);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "controllerMouseMove", function (e, _ref3) {
      var deltaX = _ref3.deltaX,
          deltaY = _ref3.deltaY;

      var isMove = Math.abs(deltaX) >= _this.openPanelPeriphery || Math.abs(deltaY) >= _this.openPanelPeriphery;

      _this.setState({
        isMove: isMove
      });
    });

    _defineProperty(_assertThisInitialized(_this), "controllerMouseUp", function (e, _ref4) {
      var x = _ref4.x,
          y = _ref4.y;

      if (_this.props.showDestroy && _this.destroyBtn && _this.destroyBtn.contains(e.target)) {
        _this.onDestroyPlayer();

        return;
      }

      if (!_this.state.isMove) {
        if (_this.state.isNeedMobileHack) {
          _this.loadAndPlayAudio();

          _this.setState({
            isNeedMobileHack: false
          });
        }

        _this.openPanel();
      }

      _this.setState({
        moveX: x,
        moveY: y
      });

      return false;
    });

    _defineProperty(_assertThisInitialized(_this), "onHandleProgress", function (value) {
      _this.audio.currentTime = value;
    });

    _defineProperty(_assertThisInitialized(_this), "onSound", function () {
      _this.setAudioVolume(_this.state.currentAudioVolume);
    });

    _defineProperty(_assertThisInitialized(_this), "setAudioVolume", function (value) {
      _this.audio.volume = value;

      _this.setState({
        currentAudioVolume: value,
        soundValue: value
      });
    });

    _defineProperty(_assertThisInitialized(_this), "stopAll", function (target) {
      target.stopPropagation();
      target.preventDefault();
    });

    _defineProperty(_assertThisInitialized(_this), "getBoundingClientRect", function (ele) {
      var _ele$getBoundingClien = ele.getBoundingClientRect(),
          left = _ele$getBoundingClien.left,
          top = _ele$getBoundingClien.top;

      return {
        left: left,
        top: top
      };
    });

    _defineProperty(_assertThisInitialized(_this), "audioLoop", function () {
      _this.setState(function (_ref5) {
        var isLoop = _ref5.isLoop;
        return {
          isLoop: !isLoop
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioReload", function () {
      _this.handlePlay(_playMode2["default"].singleLoop);

      _this.props.onAudioReload && _this.props.onAudioReload(_this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "openPanel", function () {
      if (_this.props.toggleMode) {
        _this.setState({
          toggle: true
        });

        _this.props.onModeChange && _this.props.onModeChange(_this.toggleModeName.full);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onHidePanel", function () {
      _this.setState({
        toggle: false,
        audioListsPanelVisible: false
      });

      _this.props.onModeChange && _this.props.onModeChange(_this.toggleModeName.mini);
    });

    _defineProperty(_assertThisInitialized(_this), "onDestroyPlayer", function () {
      if (_this.props.onBeforeDestroy) {
        var onBeforeDestroy = Promise.resolve(_this.props.onBeforeDestroy(_this.state.playId, _this.state.audioLists, _this.getBaseAudioInfo()));

        if (onBeforeDestroy && onBeforeDestroy.then) {
          onBeforeDestroy.then(function () {
            _this._onDestroyPlayer();
          }) // ignore unhandledrejection handler
          ["catch"](function () {});
        }

        return;
      }

      _this._onDestroyPlayer();
    });

    _defineProperty(_assertThisInitialized(_this), "_onDestroyPlayer", function () {
      _this.componentWillUnmount();

      _this.player.remove();
    });

    _defineProperty(_assertThisInitialized(_this), "_onDestroyed", function () {
      if (_this.props.onDestroyed) {
        _this.props.onDestroyed(_this.state.playId, _this.state.audioLists, _this.getBaseAudioInfo());
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onTogglePlay", function () {
      if (_this.state.audioLists.length >= 1) {
        _this.lyric.togglePlay();

        if (_this.state.playing) {
          return _this._pauseAudio();
        }

        _this.setState({
          notAutoPlayUntilPlayClicked: !_this.props.autoPlay
        }, _this.loadAndPlayAudio);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "canPlay", function () {
      _this.setAudioLength();

      _this.setState({
        loading: false,
        playing: false,
        isAudioListsChange: false
      });

      if (_this.state.isInitAutoplay) {
        _this.loadAndPlayAudio();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_pauseAudio", function () {
      _this.audio.pause();

      _this.setState({
        playing: false,
        pause: true
      }, function () {
        _this.lyric && _this.lyric.stop();
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onPauseAudio", function () {
      _this._pauseAudio();

      _this.lyric && _this.lyric.stop();
      _this.props.onAudioPause && _this.props.onAudioPause(_this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "loadAndPlayAudio", function () {
      var _this$props = _this.props,
          autoPlay = _this$props.autoPlay,
          remember = _this$props.remember;
      var _this$state3 = _this.state,
          isInitAutoplay = _this$state3.isInitAutoplay,
          isInitRemember = _this$state3.isInitRemember,
          loadProgress = _this$state3.loadProgress,
          notAutoPlayUntilPlayClicked = _this$state3.notAutoPlayUntilPlayClicked;
      var networkState = _this.audio.networkState;
      var maxLoadProgress = 100;

      _this.setState({
        loading: true
      });

      if (loadProgress < maxLoadProgress) {
        _this.setState({
          loadProgress: loadProgress + 1
        });
      }

      if (networkState !== _networkState["default"].NETWORK_NO_SOURCE) {
        var _this$getLastPlayStat = _this.getLastPlayStatus(),
            pause = _this$getLastPlayStat.pause;

        var isLastPause = remember && !isInitRemember && pause;
        var canPlay = isInitAutoplay || autoPlay || notAutoPlayUntilPlayClicked;

        _this.setState({
          playing: remember ? !isLastPause : canPlay,
          loading: false,
          pause: remember ? isLastPause : !canPlay,
          loadProgress: maxLoadProgress
        }, function () {
          if (remember ? !isLastPause : canPlay) {
            // fuck Safari is need muted :(
            // this.audio.muted = true
            _this.audio.play();
          }

          _this.setState({
            isInitAutoplay: true,
            isInitRemember: true
          });
        });
      } else {
        _this.onAudioLoadError();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "setAudioLength", function () {
      _this.setState({
        duration: _this.audio.duration
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioLoadError", function (error) {
      var _this$state4 = _this.state,
          playMode = _this$state4.playMode,
          audioLists = _this$state4.audioLists,
          playId = _this$state4.playId,
          musicSrc = _this$state4.musicSrc;
      var loadAudioErrorPlayNext = _this.props.loadAudioErrorPlayNext;
      var isSingleLoop = playMode === _playMode2["default"].singleLoop;
      var isLastAudio = (playMode === _playMode2["default"].order || playMode === _playMode2["default"].orderLoop) && playId === audioLists[audioLists.length - 1].id;
      var currentPlayMode = isSingleLoop ? _playMode2["default"].order : playMode;

      _this.lyric.stop(); //如果当前音乐加载出错 尝试播放下一首


      if (loadAudioErrorPlayNext && !isLastAudio) {
        _this.handlePlay(currentPlayMode, true);
      } // 如果删除歌曲或其他原因导致列表为空时
      // 这时候会触发 https://developer.mozilla.org/en-US/docs/Web/API/MediaError


      if (musicSrc) {
        _this.props.onAudioLoadError && _this.props.onAudioLoadError(_this.audio.error || error && error.reason || null, playId, audioLists, _this.getBaseAudioInfo());
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handlePlay", function (playMode) {
      var isNext = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      var _this$state5 = _this.state,
          playId = _this$state5.playId,
          audioLists = _this$state5.audioLists;
      var audioListsLen = audioLists.length;
      var currentPlayIndex = audioLists.findIndex(function (audio) {
        return audio.id === playId;
      });

      switch (playMode) {
        //顺序播放
        case _playMode2["default"].order:
          // 拖拽排序后 或者 正常播放 到最后一首歌 就暂停
          if (currentPlayIndex === audioListsLen - 1) return _this._pauseAudio();

          _this.audioListsPlay(isNext ? audioLists[currentPlayIndex + 1].id : audioLists[currentPlayIndex - 1].id);

          break;
        //列表循环

        case _playMode2["default"].orderLoop:
          if (isNext) {
            if (currentPlayIndex === audioListsLen - 1) {
              return _this.audioListsPlay(audioLists[0].id);
            }

            _this.audioListsPlay(audioLists[currentPlayIndex + 1].id);
          } else {
            if (currentPlayIndex === 0) {
              return _this.audioListsPlay(audioLists[audioListsLen - 1].id);
            }

            _this.audioListsPlay(audioLists[currentPlayIndex - 1].id);
          }

          break;
        //单曲循环

        case _playMode2["default"].singleLoop:
          _this.audio.currentTime = 0;

          _this.audioListsPlay(playId, true);

          break;
        //随机播放

        case _playMode2["default"].shufflePlay:
          {
            var randomIndex = (0, _utils.createRandomNum)(0, audioListsLen - 1);
            var randomPlayId = (audioLists[randomIndex] || {}).id;

            _this.audioListsPlay(randomPlayId, true);
          }
          break;

        default:
          break;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "audioEnd", function () {
      _this.props.onAudioEnded && _this.props.onAudioEnded(_this.state.playId, _this.state.audioLists, _this.getBaseAudioInfo());

      _this.handlePlay(_this.state.playMode);
    });

    _defineProperty(_assertThisInitialized(_this), "audioPrevAndNextBasePlayHandle", function () {
      var isNext = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      var playMode = _this.state.playMode;
      var _playMode = "";

      switch (playMode) {
        case _playMode2["default"].shufflePlay:
          _playMode = playMode;
          break;

        default:
          _playMode = _playMode2["default"].orderLoop;
          break;
      }

      _this.handlePlay(_playMode, isNext);
    });

    _defineProperty(_assertThisInitialized(_this), "audioPrevPlay", function () {
      _this.audioPrevAndNextBasePlayHandle(false);
    });

    _defineProperty(_assertThisInitialized(_this), "audioNextPlay", function () {
      _this.audioPrevAndNextBasePlayHandle(true);
    });

    _defineProperty(_assertThisInitialized(_this), "audioTimeUpdate", function () {
      var currentTime = _this.audio.currentTime;

      _this.setState({
        currentTime: currentTime
      });

      if (_this.props.remember) {
        _this.saveLastPlayStatus();
      }

      _this.props.onAudioProgress && _this.props.onAudioProgress(_this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "audioSoundChange", function (value) {
      _this.setAudioVolume(value);
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioVolumeChange", function () {
      var volume = _this.audio.volume;

      _this.setState({
        isMute: _this.audio.volume <= 0,
        soundValue: volume
      });

      _this.props.onAudioVolumeChange && _this.props.onAudioVolumeChange(volume);
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioPlay", function () {
      _this.setState({
        playing: true,
        loading: false
      }, _this.initLyricParser);

      _this.props.onAudioPlay && _this.props.onAudioPlay(_this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioSeeked", function () {
      if (_this.state.audioLists.length >= 1) {
        if (_this.state.playing) {
          _this.setState({
            playing: true
          }, function () {
            _this.loadAndPlayAudio();

            _this.lyric.seek(_this.getLyricPlayTime());
          });
        }

        _this.props.onAudioSeeked && _this.props.onAudioSeeked(_this.getBaseAudioInfo());
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onMute", function () {
      _this.setState({
        isMute: true,
        soundValue: 0,
        currentAudioVolume: _this.audio.volume
      }, function () {
        _this.audio.volume = 0;
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onAudioAbort", function (e) {
      var _this$state6 = _this.state,
          audioLists = _this$state6.audioLists,
          playId = _this$state6.playId;

      var audioInfo = _this.getBaseAudioInfo();

      var mergedAudioInfo = Object.assign({}, e, audioInfo);
      _this.props.onAudioAbort && _this.props.onAudioAbort(playId, audioLists, mergedAudioInfo);

      if (audioLists.length) {
        _this.audio.pause();

        _this.state.isInitAutoplay && _this.audio.play();

        _this.lyric.stop();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "toggleMode", function (mode) {
      if (mode === _this.toggleModeName["full"]) {
        _this.setState({
          toggle: true
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "audioListsDragEnd", function (fromIndex, toIndex) {
      var _this$state7 = _this.state,
          playId = _this$state7.playId,
          audioLists = _this$state7.audioLists;

      var _audioLists = _toConsumableArray(audioLists);

      var item = _audioLists.splice(fromIndex, 1)[0];

      _audioLists.splice(toIndex, 0, item); //如果拖动正在播放的歌曲 播放Id 等于 拖动后的index


      var _playId = fromIndex === playId ? toIndex : playId;

      _this.setState({
        audioLists: _audioLists,
        playId: _playId
      });

      _this.props.onAudioListsDragEnd && _this.props.onAudioListsDragEnd(fromIndex, toIndex);
      _this.props.onAudioListsChange && _this.props.onAudioListsChange(_playId, _audioLists, _this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "saveLastPlayStatus", function () {
      var _this$state8 = _this.state,
          currentTime = _this$state8.currentTime,
          playId = _this$state8.playId,
          duration = _this$state8.duration,
          theme = _this$state8.theme,
          soundValue = _this$state8.soundValue,
          playMode = _this$state8.playMode,
          name = _this$state8.name,
          cover = _this$state8.cover,
          singer = _this$state8.singer,
          musicSrc = _this$state8.musicSrc,
          pause = _this$state8.pause;
      var lastPlayStatus = JSON.stringify({
        currentTime: currentTime,
        playId: playId,
        duration: duration,
        theme: theme,
        playMode: playMode,
        soundValue: soundValue,
        name: name,
        cover: cover,
        singer: singer,
        musicSrc: musicSrc,
        pause: pause
      });
      localStorage.setItem("lastPlayStatus", lastPlayStatus);
    });

    _defineProperty(_assertThisInitialized(_this), "getLastPlayStatus", function () {
      var _this$props2 = _this.props,
          theme = _this$props2.theme,
          defaultPlayMode = _this$props2.defaultPlayMode;
      var status = {
        currentTime: 0,
        duration: 0,
        playMode: defaultPlayMode,
        name: "",
        cover: "",
        singer: "",
        musicSrc: "",
        lyric: "",
        playId: _this.getDefaultPlayId(),
        theme: theme,
        pause: false
      };

      try {
        return JSON.parse(localStorage.getItem("lastPlayStatus")) || status;
      } catch (error) {
        return status;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "mockAutoPlayForMobile", function () {
      if (_this.props.autoPlay && !_this.state.playing && _this.state.pause) {
        _this.audio.load();

        _this.audio.pause();

        _this.audio.play();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "bindMobileAutoPlayEvents", function () {
      document.addEventListener("touchstart", function () {
        _this.mockAutoPlayForMobile();
      }, {
        once: true
      }); //监听微信准备就绪事件

      document.addEventListener("WeixinJSBridgeReady", function () {
        _this.mockAutoPlayForMobile();
      });
    });

    _defineProperty(_assertThisInitialized(_this), "bindSafariAutoPlayEvents", function () {
      document.addEventListener("click", function () {
        _this.mockAutoPlayForMobile();
      }, {
        once: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "unBindEvents", function () {
      var _this2;

      (_this2 = _this).bindEvents.apply(_this2, arguments);
    });

    _defineProperty(_assertThisInitialized(_this), "bindEvents", function () {
      var target = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _this.audio;
      var eventsNames = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
        waiting: _this.loadAndPlayAudio,
        canplay: _this.canPlay,
        error: _this.onAudioLoadError,
        ended: _this.audioEnd,
        seeked: _this.onAudioSeeked,
        pause: _this.onPauseAudio,
        play: _this.onAudioPlay,
        timeupdate: _this.audioTimeUpdate,
        volumechange: _this.onAudioVolumeChange,
        stalled: _this.onAudioLoadError,
        //当浏览器尝试获取媒体数据，但数据不可用时
        abort: _this.onAudioAbort
      };
      var bind = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
      var once = _this.props.once;

      for (var name in eventsNames) {
        var _events = eventsNames[name];
        bind ? target.addEventListener(name, _events, {
          once: !!(once && name === "play")
        }) : target.removeEventListener(name, _events);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getPlayId", function () {
      var audioLists = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _this.state.audioLists;
      var playIndex = Math.max(0, Math.min(audioLists.length, _this.props.defaultPlayIndex));
      var playId = _this.state.playId || audioLists[playIndex].id;
      return playId;
    });

    _defineProperty(_assertThisInitialized(_this), "getPlayInfo", function () {
      var audioLists = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
      var newAudioLists = audioLists.filter(function (audio) {
        return !audio["id"];
      });
      var lastAudioLists = audioLists.filter(function (audio) {
        return audio["id"];
      });
      var mergedAudioLists = [].concat(_toConsumableArray(lastAudioLists), _toConsumableArray(newAudioLists.map(function (info) {
        return _objectSpread({}, info, {
          id: (0, _utils.uuId)()
        });
      })));

      var playId = _this.getPlayId(mergedAudioLists);

      var _ref6 = mergedAudioLists.find(function (_ref7) {
        var id = _ref7.id;
        return id === playId;
      }) || {},
          _ref6$name = _ref6.name,
          name = _ref6$name === void 0 ? "" : _ref6$name,
          _ref6$cover = _ref6.cover,
          cover = _ref6$cover === void 0 ? "" : _ref6$cover,
          _ref6$singer = _ref6.singer,
          singer = _ref6$singer === void 0 ? "" : _ref6$singer,
          _ref6$musicSrc = _ref6.musicSrc,
          musicSrc = _ref6$musicSrc === void 0 ? "" : _ref6$musicSrc,
          _ref6$lyric = _ref6.lyric,
          lyric = _ref6$lyric === void 0 ? "" : _ref6$lyric;

      return {
        name: name,
        cover: cover,
        singer: singer,
        musicSrc: musicSrc,
        lyric: lyric,
        audioLists: mergedAudioLists,
        playId: playId
      };
    });

    _defineProperty(_assertThisInitialized(_this), "getPlayInfoOfNewList", function () {
      var audioLists = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

      var _audioLists = audioLists.map(function (info) {
        return _objectSpread({}, info, {
          id: (0, _utils.uuId)()
        });
      });

      var playIndex = Math.max(0, Math.min(_audioLists.length, _this.props.defaultPlayIndex));
      var playId = _this.state.playId || _audioLists[playIndex].id;

      var _ref8 = _audioLists.find(function (_ref9) {
        var id = _ref9.id;
        return id === playId;
      }) || {},
          _ref8$name = _ref8.name,
          name = _ref8$name === void 0 ? "" : _ref8$name,
          _ref8$cover = _ref8.cover,
          cover = _ref8$cover === void 0 ? "" : _ref8$cover,
          _ref8$singer = _ref8.singer,
          singer = _ref8$singer === void 0 ? "" : _ref8$singer,
          _ref8$musicSrc = _ref8.musicSrc,
          musicSrc = _ref8$musicSrc === void 0 ? "" : _ref8$musicSrc,
          _ref8$lyric = _ref8.lyric,
          lyric = _ref8$lyric === void 0 ? "" : _ref8$lyric;

      return {
        name: name,
        cover: cover,
        singer: singer,
        musicSrc: musicSrc,
        lyric: lyric,
        audioLists: _audioLists,
        playId: playId
      };
    });

    _defineProperty(_assertThisInitialized(_this), "initPlayInfo", function (audioLists, cb) {
      var info = _this.getPlayInfo(audioLists);

      switch (_typeof(info.musicSrc)) {
        case "function":
          info.musicSrc().then(function (originMusicSrc) {
            _this.setState(_objectSpread({}, info, {
              musicSrc: originMusicSrc
            }), cb);
          }, _this.onAudioLoadError);
          break;

        default:
          _this.setState(info, cb);

      }
    });

    _defineProperty(_assertThisInitialized(_this), "listenerIsMobile", function (_ref10) {
      var matches = _ref10.matches;

      _this.setState({
        isMobile: !!matches
      });
    });

    _defineProperty(_assertThisInitialized(_this), "addMobileListener", function () {
      _this.media = window.matchMedia("(max-width: 768px) and (orientation : portrait)");

      _this.media.addListener(_this.listenerIsMobile);
    });

    _defineProperty(_assertThisInitialized(_this), "setDefaultAudioVolume", function () {
      var _this$props3 = _this.props,
          defaultVolume = _this$props3.defaultVolume,
          remember = _this$props3.remember; //音量 [0-1]

      _this.defaultVolume = Math.max(0, Math.min(defaultVolume, 1));

      var _this$getLastPlayStat2 = _this.getLastPlayStatus(),
          _this$getLastPlayStat3 = _this$getLastPlayStat2.soundValue,
          soundValue = _this$getLastPlayStat3 === void 0 ? _this.defaultVolume : _this$getLastPlayStat3;

      _this.setAudioVolume(remember ? soundValue : _this.defaultVolume);
    });

    _defineProperty(_assertThisInitialized(_this), "getDefaultPlayId", function () {
      var audioLists = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _this.props.audioLists;
      var playIndex = Math.max(0, Math.min(audioLists.length, _this.props.defaultPlayIndex));
      return audioLists[playIndex] && audioLists[playIndex].id;
    });

    _defineProperty(_assertThisInitialized(_this), "getLyricPlayTime", function () {
      var _formatTime$split = (0, _utils.formatTime)(_this.audio.currentTime).split(":"),
          _formatTime$split2 = _slicedToArray(_formatTime$split, 2),
          m = _formatTime$split2[0],
          s = _formatTime$split2[1];

      return m * 1000 + s * 10;
    });

    _defineProperty(_assertThisInitialized(_this), "initLyricParser", function () {
      _this.lyric = undefined;
      _this.lyric = new _lyric["default"](_this.state.lyric, _this.onLyricChange);

      _this.lyric.stop();

      if (_this.props.showLyric && _this.state.playing) {
        _this.lyric.seek(_this.getLyricPlayTime());

        _this.lyric.play();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onLyricChange", function (_ref11) {
      var lineNum = _ref11.lineNum,
          txt = _ref11.txt;

      _this.setState({
        currentLyric: txt
      });

      _this.props.onAudioLyricChange && _this.props.onAudioLyricChange(lineNum, txt);
    });

    _defineProperty(_assertThisInitialized(_this), "updateTheme", function (theme) {
      if (theme && theme !== _this.props.theme && (theme === _this.lightThemeName || theme === _this.darkThemeName)) {
        _this.setState({
          theme: theme
        });

        _this.props.onThemeChange && _this.props.onThemeChange(theme);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "updateMode", function (mode) {
      if (mode && mode !== _this.props.mode && (mode === _this.toggleModeName.full || mode === _this.toggleModeName.mini)) {
        _this.setState({
          toggle: mode === _this.toggleModeName.full
        });

        _this.props.onModeChange && _this.props.onModeChange(mode);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "updateAudioLists", function (audioLists) {
      var newAudioLists = [].concat(_toConsumableArray(_this.state.audioLists), _toConsumableArray(audioLists.filter(function (audio) {
        return _this.state.audioLists.findIndex(function (v) {
          return v.musicSrc === audio.musicSrc;
        }) === -1;
      })));

      _this.initPlayInfo(newAudioLists);

      _this.bindEvents(_this.audio);

      _this.props.onAudioListsChange && _this.props.onAudioListsChange(_this.state.playId, audioLists, _this.getBaseAudioInfo());
    });

    _defineProperty(_assertThisInitialized(_this), "loadNewAudioLists", function (audioLists, _ref12) {
      var remember = _ref12.remember,
          defaultPlayIndex = _ref12.defaultPlayIndex,
          defaultPlayMode = _ref12.defaultPlayMode,
          theme = _ref12.theme,
          autoPlayInitLoadPlayList = _ref12.autoPlayInitLoadPlayList;

      if (audioLists.length >= 1) {
        var info = _this.getPlayInfoOfNewList(audioLists);

        var lastPlayStatus = remember ? _this.getLastPlayStatus(defaultPlayIndex) : {
          playMode: defaultPlayMode,
          theme: theme
        };

        switch (_typeof(info.musicSrc)) {
          case "function":
            info.musicSrc().then(function (val) {
              _this.setState(_objectSpread({}, info, {
                musicSrc: val,
                isInitAutoplay: autoPlayInitLoadPlayList
              }, lastPlayStatus));
            }, _this.onAudioLoadError);
            break;

          default:
            _this.setState(_objectSpread({}, info, {
              isInitAutoplay: autoPlayInitLoadPlayList
            }, lastPlayStatus));

        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "changeAudioLists", function (audioLists) {
      _this.resetAudioStatus();

      _this.loadNewAudioLists(audioLists, _this.props);

      _this.props.onAudioListsChange && _this.props.onAudioListsChange(_this.state.playId, audioLists, _this.getBaseAudioInfo());

      _this.setState({
        isAudioListsChange: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "resetPlayList", function (state) {
      var _playIndex = Math.max(0, Math.min(state.audioLists.length, _this.props.defaultPlayIndex));

      var currentPlay = state.audioLists[_playIndex];

      if (currentPlay && currentPlay.id) {
        _this.audioListsPlay(currentPlay.id, true, state);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "updatePlayIndex", function (playIndex) {
      // 播放索引 改变
      var currentPlayIndex = _this.state.audioLists.findIndex(function (audio) {
        return audio.id === _this.state.playId;
      });

      if (currentPlayIndex !== playIndex) {
        var _playIndex = Math.max(0, Math.min(_this.state.audioLists.length, playIndex));

        var currentPlay = _this.state.audioLists[_playIndex];

        if (currentPlay && currentPlay.id) {
          _this.audioListsPlay(currentPlay.id, true);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onGetAudioInstance", function () {
      if (_this.props.getAudioInstance) {
        Object.defineProperty(_this.audio, "destroy", {
          value: function value() {
            _this.onDestroyPlayer();
          },
          writable: false
        });

        _this.props.getAudioInstance(_this.audio);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "updateMediaSessionMetadata", function () {
      if ("mediaSession" in navigator && _this.props.showMediaSession) {
        var _this$state9 = _this.state,
            name = _this$state9.name,
            cover = _this$state9.cover,
            singer = _this$state9.singer;
        navigator.mediaSession.metadata = new MediaMetadata({
          title: name,
          artist: singer,
          album: name,
          artwork: [{
            src: cover,
            sizes: "96x96",
            type: "image/png"
          }, {
            src: cover,
            sizes: "128x128",
            type: "image/png"
          }, {
            src: cover,
            sizes: "192x192",
            type: "image/png"
          }, {
            src: cover,
            sizes: "256x256",
            type: "image/png"
          }, {
            src: cover,
            sizes: "384x384",
            type: "image/png"
          }, {
            src: cover,
            sizes: "512x512",
            type: "image/png"
          }]
        });

        _this.updateMediaSessionPositionState();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "updateMediaSessionPositionState", function () {
      if ("setPositionState" in navigator.mediaSession) {
        try {
          var _assertThisInitialize = _assertThisInitialized(_this),
              audio = _assertThisInitialize.audio;

          navigator.mediaSession.setPositionState({
            duration: audio.duration,
            playbackRate: audio.playbackRate,
            position: audio.currentTime
          });
        } catch (error) {
          console.log("setPositionState error: ", error);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onAddMediaSession", function () {
      if ("mediaSession" in navigator && _this.props.showMediaSession) {
        var defaultSkipTime = 10;

        _this.updateMediaSessionMetadata();

        navigator.mediaSession.setActionHandler("play", _this.onTogglePlay);
        navigator.mediaSession.setActionHandler("pause", _this.onTogglePlay);
        navigator.mediaSession.setActionHandler("seekbackward", function (details) {
          var skipTime = details.seekOffset || defaultSkipTime;
          _this.audio.currentTime = Math.max(_this.audio.currentTime - skipTime, 0);
        });
        navigator.mediaSession.setActionHandler("seekforward", function (details) {
          var skipTime = details.seekOffset || defaultSkipTime;
          _this.audio.currentTime = Math.min(_this.audio.currentTime + skipTime, _this.audio.duration);
        });
        navigator.mediaSession.setActionHandler("previoustrack", _this.audioPrevPlay);
        navigator.mediaSession.setActionHandler("nexttrack", _this.audioNextPlay);

        try {
          navigator.mediaSession.setActionHandler("seekto", function (event) {
            console.log('> User clicked "Seek To" icon.');

            if (event.fastSeek && "fastSeek" in _this.audio) {
              _this.audio.fastSeek(event.seekTime);

              return;
            }

            _this.audio.currentTime = event.seekTime;

            _this.updateMediaSessionPositionState();
          });
        } catch (error) {
          console.log('Warning! The "seekto" media session action is not supported.');
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "bindUnhandledRejection", function () {
      window.addEventListener("unhandledrejection", _this.onAudioLoadError);
    });

    _defineProperty(_assertThisInitialized(_this), "unBindUnhandledRejection", function () {
      window.removeEventListener("unhandledrejection", _this.onAudioLoadError);
    });

    _defineProperty(_assertThisInitialized(_this), "bindKeyDownEvents", function () {
      document.addEventListener("keydown", _this.onKeyDown, false);
    });

    _defineProperty(_assertThisInitialized(_this), "unBindKeyDownEvents", function () {
      document.removeEventListener("keydown", _this.onKeyDown, false);
    });

    _defineProperty(_assertThisInitialized(_this), "onKeyDown", function (e) {
      var spaceBar = _this.props.spaceBar;

      if (spaceBar && e.keyCode === _keycode.SPACE_BAR_KEYCODE) {
        _this.onTogglePlay();
      }
    });

    _this.audio = null; //当前播放器

    _this.lightThemeName = "light";
    _this.darkThemeName = "dark"; //模式配置

    _this.toggleModeName = {
      full: "full",
      mini: "mini"
    };
    _this.targetId = "music-player-controller";
    _this.openPanelPeriphery = 1; //移动差值 在 这之间 认为是点击打开panel
    //播放模式配置

    _this._PLAY_MODE_ = Object.values(_playMode2["default"]);
    _this._PLAY_MODE_LENGTH_ = _this._PLAY_MODE_.length;
    return _this;
  }

  _createClass(ReactJkMusicPlayer, [{
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props4 = this.props,
          className = _this$props4.className,
          drag = _this$props4.drag,
          style = _this$props4.style,
          showDownload = _this$props4.showDownload,
          showPlay = _this$props4.showPlay,
          showReload = _this$props4.showReload,
          showPlayMode = _this$props4.showPlayMode,
          showThemeSwitch = _this$props4.showThemeSwitch,
          toggleMode = _this$props4.toggleMode,
          showMiniModeCover = _this$props4.showMiniModeCover,
          extendsContent = _this$props4.extendsContent,
          defaultPlayMode = _this$props4.defaultPlayMode,
          seeked = _this$props4.seeked,
          showProgressLoadBar = _this$props4.showProgressLoadBar,
          bounds = _this$props4.bounds,
          defaultPosition = _this$props4.defaultPosition,
          showMiniProcessBar = _this$props4.showMiniProcessBar,
          preload = _this$props4.preload,
          glassBg = _this$props4.glassBg,
          remove = _this$props4.remove,
          lyricClassName = _this$props4.lyricClassName,
          showLyric = _this$props4.showLyric,
          emptyLyricText = _this$props4.emptyLyricText,
          getContainer = _this$props4.getContainer,
          autoHiddenCover = _this$props4.autoHiddenCover,
          showDestroy = _this$props4.showDestroy;
      var locale = this.locale;
      var _this$state10 = this.state,
          toggle = _this$state10.toggle,
          playing = _this$state10.playing,
          duration = _this$state10.duration,
          currentTime = _this$state10.currentTime,
          isMute = _this$state10.isMute,
          soundValue = _this$state10.soundValue,
          moveX = _this$state10.moveX,
          moveY = _this$state10.moveY,
          loading = _this$state10.loading,
          audioListsPanelVisible = _this$state10.audioListsPanelVisible,
          pause = _this$state10.pause,
          theme = _this$state10.theme,
          name = _this$state10.name,
          cover = _this$state10.cover,
          singer = _this$state10.singer,
          musicSrc = _this$state10.musicSrc,
          playId = _this$state10.playId,
          isMobile = _this$state10.isMobile,
          playMode = _this$state10.playMode,
          playModeTipVisible = _this$state10.playModeTipVisible,
          playModelNameVisible = _this$state10.playModelNameVisible,
          initAnimate = _this$state10.initAnimate,
          loadProgress = _this$state10.loadProgress,
          audioLists = _this$state10.audioLists,
          removeId = _this$state10.removeId,
          currentLyric = _this$state10.currentLyric,
          audioLyricVisible = _this$state10.audioLyricVisible;
      var preloadState = preload === false || preload === "none" ? {} : preload === true ? {
        preload: "auto"
      } : {
        preload: preload
      };
      var panelToggleAnimate = initAnimate ? {
        show: audioListsPanelVisible,
        hide: !audioListsPanelVisible
      } : {
        show: audioListsPanelVisible
      };
      var currentPlayMode = _playMode2["default"][playMode || defaultPlayMode] || _playMode2["default"].order;
      var currentPlayModeName = locale.playModeText[currentPlayMode];
      var isShowMiniModeCover = showMiniModeCover && !autoHiddenCover || autoHiddenCover && cover ? {
        style: {
          backgroundImage: "url(".concat(cover, ")")
        }
      } : {};

      var _currentTime = (0, _utils.formatTime)(currentTime);

      var _duration = (0, _utils.formatTime)(duration);

      var progressHandler = seeked ? {
        onChange: this.onHandleProgress,
        onAfterChange: this.onAudioSeeked
      } : {}; //进度条

      var ProgressBar = /*#__PURE__*/_react["default"].createElement(_Slider["default"], _extends({
        max: Math.ceil(duration),
        defaultValue: 0,
        value: Math.ceil(currentTime)
      }, progressHandler, _slider.sliderBaseOptions)); //下载按钮


      var DownloadComponent = showDownload && /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group audio-download"
      }, _defineProperty({}, IS_MOBILE ? "onTouchStart" : "onClick", this.onAudioDownload), {
        title: locale.downloadText
      }), /*#__PURE__*/_react["default"].createElement(_Icon.DownloadIcon, null)); //主题开关


      var ThemeSwitchComponent = showThemeSwitch && /*#__PURE__*/_react["default"].createElement("span", {
        className: "group theme-switch"
      }, /*#__PURE__*/_react["default"].createElement(_rcSwitch["default"], {
        className: "theme-switch-container",
        onChange: this.themeChange,
        checkedChildren: locale.lightThemeText,
        unCheckedChildren: locale.darkThemeText,
        checked: theme === this.lightThemeName,
        title: locale.switchThemeText
      })); //重放


      var ReloadComponent = showReload && /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group reload-btn"
      }, IS_MOBILE ? {
        onTouchStart: this.onAudioReload
      } : {
        onClick: this.onAudioReload
      }, {
        title: locale.reloadText
      }), /*#__PURE__*/_react["default"].createElement(_Icon.ReloadIcon, null)); //歌词


      var LyricComponent = showLyric && /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: (0, _classnames["default"])("group lyric-btn", {
          "lyric-btn-active": audioLyricVisible
        })
      }, IS_MOBILE ? {
        onTouchStart: this.toggleAudioLyric
      } : {
        onClick: this.toggleAudioLyric
      }, {
        title: locale.toggleLyricText
      }), /*#__PURE__*/_react["default"].createElement(_Icon.LyricIcon, null)); //播放模式


      var PlayModeComponent = showPlayMode && /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: (0, _classnames["default"])("group loop-btn")
      }, IS_MOBILE ? {
        onTouchStart: this.togglePlayMode
      } : {
        onClick: this.togglePlayMode
      }, {
        title: locale.playModeText[currentPlayMode]
      }), this.renderPlayModeIcon(currentPlayMode));

      var miniProcessBarR = isMobile ? 30 : 40;

      var DestroyComponent = showDestroy && /*#__PURE__*/_react["default"].createElement("span", _extends({
        title: locale.destroyText,
        className: "group destroy-btn",
        ref: function ref(node) {
          return _this3.destroyBtn = node;
        }
      }, !drag || toggle ? _defineProperty({}, IS_MOBILE ? "onTouchStart" : "onClick", this.onDestroyPlayer) : null), /*#__PURE__*/_react["default"].createElement(_Icon.CloseIcon, null));

      var AudioController = /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("react-jinke-music-player"),
        style: defaultPosition
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("music-player")
      }, showMiniProcessBar && /*#__PURE__*/_react["default"].createElement(_CircleProcessBar["default"], {
        progress: currentTime / duration,
        r: miniProcessBarR
      }), /*#__PURE__*/_react["default"].createElement("div", _extends({
        id: this.targetId,
        className: (0, _classnames["default"])("scale", "music-player-controller", {
          "music-player-playing": this.state.playing
        })
      }, isShowMiniModeCover), loading ? /*#__PURE__*/_react["default"].createElement(_Load["default"], null) : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("span", {
        className: "controller-title"
      }, locale.controllerTitle), /*#__PURE__*/_react["default"].createElement("div", {
        className: "music-player-controller-setting"
      }, toggle ? locale.closeText : locale.openText))), DestroyComponent));

      var container = getContainer() || document.body;
      var audioTitle = this.getAudioTitle();
      return (0, _reactDom.createPortal)( /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("react-jinke-music-player-main", {
          "light-theme": theme === this.lightThemeName,
          "dark-theme": theme === this.darkThemeName
        }, className),
        style: style,
        ref: function ref(player) {
          return _this3.player = player;
        }
      }, toggle && isMobile && /*#__PURE__*/_react["default"].createElement(_PlayerMobile["default"], {
        playing: playing,
        loading: loading,
        pause: pause,
        name: name,
        singer: singer,
        cover: cover,
        themeSwitch: ThemeSwitchComponent,
        duration: _duration,
        currentTime: _currentTime,
        progressBar: ProgressBar,
        onPlay: this.onTogglePlay,
        currentPlayModeName: currentPlayModeName,
        playMode: PlayModeComponent,
        audioNextPlay: this.audioNextPlay,
        audioPrevPlay: this.audioPrevPlay,
        playListsIcon: /*#__PURE__*/_react["default"].createElement(_Icon.PlayLists, null),
        reloadIcon: ReloadComponent,
        downloadIcon: DownloadComponent,
        nextAudioIcon: /*#__PURE__*/_react["default"].createElement(_Icon.NextAudioIcon, null),
        prevAudioIcon: /*#__PURE__*/_react["default"].createElement(_Icon.PrevAudioIcon, null),
        playIcon: /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePlayIcon, null),
        pauseIcon: /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePauseIcon, null),
        closeIcon: /*#__PURE__*/_react["default"].createElement(_Icon.CloseIcon, null),
        loadingIcon: /*#__PURE__*/_react["default"].createElement(_Load["default"], null),
        playModeTipVisible: playModeTipVisible,
        openAudioListsPanel: this.openAudioListsPanel,
        onClose: this.onHidePanel,
        extendsContent: extendsContent,
        glassBg: glassBg,
        LyricIcon: LyricComponent,
        autoHiddenCover: autoHiddenCover
      }), toggle ? undefined : drag ? /*#__PURE__*/_react["default"].createElement(_reactDraggable["default"], {
        bounds: bounds,
        position: {
          x: moveX,
          y: moveY
        },
        onDrag: this.controllerMouseMove,
        onStop: this.controllerMouseUp,
        onStart: this.controllerMouseMove
      }, AudioController) : AudioController, toggle && !isMobile && /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("music-player-panel", "translate", {
          "glass-bg": glassBg
        })
      }, /*#__PURE__*/_react["default"].createElement("section", {
        className: "panel-content"
      }, (!autoHiddenCover || autoHiddenCover && cover) && /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("img-content", "img-rotate", {
          "img-rotate-pause": pause || !playing || !cover
        }),
        style: {
          backgroundImage: "url(".concat(cover, ")")
        }
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: "progress-bar-content"
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: "audio-title"
      }, audioTitle), /*#__PURE__*/_react["default"].createElement("section", {
        className: "audio-main"
      }, /*#__PURE__*/_react["default"].createElement("span", {
        className: "current-time"
      }, loading ? "--" : _currentTime), /*#__PURE__*/_react["default"].createElement("div", {
        className: "progress-bar"
      }, showProgressLoadBar && /*#__PURE__*/_react["default"].createElement("div", {
        className: "progress-load-bar",
        style: {
          width: "".concat(Math.min(loadProgress, 100), "%")
        }
      }), ProgressBar), /*#__PURE__*/_react["default"].createElement("span", {
        className: "duration"
      }, loading ? "--" : _duration))), /*#__PURE__*/_react["default"].createElement("div", {
        className: "player-content"
      }, loading ? /*#__PURE__*/_react["default"].createElement(_Load["default"], null) : showPlay ? /*#__PURE__*/_react["default"].createElement("span", {
        className: "group"
      }, /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group prev-audio",
        title: locale.previousTrackText
      }, IS_MOBILE ? {
        onTouchStart: this.audioPrevPlay
      } : {
        onClick: this.audioPrevPlay
      }), /*#__PURE__*/_react["default"].createElement(_Icon.PrevAudioIcon, null)), /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group play-btn",
        ref: function ref(node) {
          return _this3.playBtn = node;
        }
      }, IS_MOBILE ? {
        onTouchStart: this.onTogglePlay
      } : {
        onClick: this.onTogglePlay
      }, {
        title: playing ? locale.clickToPauseText : locale.clickToPlayText
      }), playing ? /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePauseIcon, null)) : /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePlayIcon, null))), /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group next-audio",
        title: locale.nextTrackText
      }, IS_MOBILE ? {
        onTouchStart: this.audioNextPlay
      } : {
        onClick: this.audioNextPlay
      }), /*#__PURE__*/_react["default"].createElement(_Icon.NextAudioIcon, null))) : undefined, ReloadComponent, DownloadComponent, ThemeSwitchComponent, extendsContent || null, /*#__PURE__*/_react["default"].createElement("span", {
        className: "group play-sounds",
        title: locale.volumeText
      }, isMute ? /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "sounds-icon"
      }, IS_MOBILE ? {
        onTouchStart: this.onSound
      } : {
        onClick: this.onSound
      }), /*#__PURE__*/_react["default"].createElement(_Icon.MdVolumeMuteIcon, null)) : /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "sounds-icon"
      }, IS_MOBILE ? {
        onTouchStart: this.onMute
      } : {
        onClick: this.onMute
      }), /*#__PURE__*/_react["default"].createElement(_Icon.MdVolumeDownIcon, null)), /*#__PURE__*/_react["default"].createElement(_Slider["default"], _extends({
        max: 1,
        value: soundValue,
        onChange: this.audioSoundChange,
        className: "sound-operation"
      }, _slider.sliderBaseOptions))), PlayModeComponent, LyricComponent, /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group audio-lists-btn",
        title: locale.playListsText
      }, IS_MOBILE ? {
        onTouchStart: this.openAudioListsPanel
      } : {
        onClick: this.openAudioListsPanel
      }), /*#__PURE__*/_react["default"].createElement("span", {
        className: "audio-lists-icon"
      }, /*#__PURE__*/_react["default"].createElement(_Icon.PlayLists, null)), /*#__PURE__*/_react["default"].createElement("span", {
        className: "audio-lists-num"
      }, audioLists.length)), toggleMode && /*#__PURE__*/_react["default"].createElement("span", _extends({
        className: "group hide-panel",
        title: locale.toggleMiniModeText
      }, IS_MOBILE ? {
        onTouchStart: this.onHidePanel
      } : {
        onClick: this.onHidePanel
      }), /*#__PURE__*/_react["default"].createElement(_Icon.FaMinusSquareOIcon, null)), DestroyComponent)), /*#__PURE__*/_react["default"].createElement(_PlayModel["default"], {
        visible: playModelNameVisible,
        value: currentPlayModeName
      })), /*#__PURE__*/_react["default"].createElement(_AudioListsPanel["default"], {
        playId: playId,
        pause: pause,
        loading: loading ? /*#__PURE__*/_react["default"].createElement(_Load["default"], null) : undefined,
        visible: audioListsPanelVisible,
        audioLists: audioLists,
        onPlay: this.audioListsPlay,
        onCancel: this.closeAudioListsPanel,
        playIcon: /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePlayIcon, null),
        pauseIcon: /*#__PURE__*/_react["default"].createElement(_Icon.AnimatePauseIcon, null),
        closeIcon: /*#__PURE__*/_react["default"].createElement(_Icon.CloseIcon, null),
        isMobile: isMobile,
        panelToggleAnimate: panelToggleAnimate,
        glassBg: glassBg,
        cover: cover,
        remove: remove,
        deleteIcon: /*#__PURE__*/_react["default"].createElement(_Icon.DeleteIcon, null),
        onDelete: this.deleteAudioLists,
        removeId: removeId,
        audioListsDragEnd: this.audioListsDragEnd,
        locale: locale
      }), audioLyricVisible && /*#__PURE__*/_react["default"].createElement(_reactDraggable["default"], null, /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("music-player-lyric", lyricClassName)
      }, currentLyric || emptyLyricText)), /*#__PURE__*/_react["default"].createElement("audio", _extends({
        className: "music-player-audio",
        title: audioTitle
      }, preloadState, {
        src: musicSrc,
        ref: function ref(node) {
          return _this3.audio = node;
        }
      }))), container);
    }
  }, {
    key: "getBaseAudioInfo",
    //返回给使用者的 音乐信息
    value: function getBaseAudioInfo() {
      var _this$state11 = this.state,
          playId = _this$state11.playId,
          cover = _this$state11.cover,
          name = _this$state11.name,
          musicSrc = _this$state11.musicSrc,
          soundValue = _this$state11.soundValue,
          lyric = _this$state11.lyric,
          audioLists = _this$state11.audioLists;
      var _this$audio = this.audio,
          currentTime = _this$audio.currentTime,
          duration = _this$audio.duration,
          muted = _this$audio.muted,
          networkState = _this$audio.networkState,
          readyState = _this$audio.readyState,
          played = _this$audio.played,
          paused = _this$audio.paused,
          ended = _this$audio.ended,
          startDate = _this$audio.startDate;
      var currentPlayIndex = audioLists.findIndex(function (audio) {
        return audio.id === playId;
      });
      var currentAudioListInfo = audioLists[currentPlayIndex] || {};
      return _objectSpread({}, currentAudioListInfo, {
        cover: cover,
        name: name,
        musicSrc: musicSrc,
        volume: soundValue,
        currentTime: currentTime,
        duration: duration,
        muted: muted,
        networkState: networkState,
        readyState: readyState,
        played: played,
        paused: paused,
        ended: ended,
        startDate: startDate,
        lyric: lyric
      });
    } //播放

  }, {
    key: "UNSAFE_componentWillReceiveProps",
    //当父组件 更新 props 时 如 audioLists 改变 更新播放信息
    // eslint-disable-next-line camelcase
    value: function UNSAFE_componentWillReceiveProps(_ref15) {
      var audioLists = _ref15.audioLists,
          playIndex = _ref15.playIndex,
          theme = _ref15.theme,
          mode = _ref15.mode,
          clearPriorAudioLists = _ref15.clearPriorAudioLists;

      if (!(0, _utils.arrayEqual)(audioLists)(this.props.audioLists)) {
        if (clearPriorAudioLists) {
          this.changeAudioLists(audioLists);
        } else {
          this.updateAudioLists(audioLists);
        }
      } else {
        this.updatePlayIndex(playIndex);
      }

      this.updateTheme(theme);
      this.updateMode(mode);
    } // eslint-disable-next-line camelcase

  }, {
    key: "UNSAFE_componentWillUpdate",
    value: function UNSAFE_componentWillUpdate(nextProps, nextState) {
      if (nextProps.clearPriorAudioLists) {
        if (nextState.isAudioListsChange !== this.state.isAudioListsChange) {
          this.resetPlayList(nextState);
        }
      }
    } // eslint-disable-next-line camelcase

  }, {
    key: "UNSAFE_componentWillMount",
    value: function UNSAFE_componentWillMount() {
      var _this4 = this;

      var _this$props5 = this.props,
          mode = _this$props5.mode,
          audioLists = _this$props5.audioLists,
          defaultPlayMode = _this$props5.defaultPlayMode,
          remember = _this$props5.remember,
          theme = _this$props5.theme,
          defaultPlayIndex = _this$props5.defaultPlayIndex,
          autoPlay = _this$props5.autoPlay; //切换 'mini' 或者 'full' 模式

      this.toggleMode(mode);

      if (audioLists.length >= 1) {
        var info = _objectSpread({}, this.getPlayInfo(audioLists), {
          isInitAutoplay: autoPlay
        });

        var lastPlayStatus = remember ? this.getLastPlayStatus(defaultPlayIndex) : {
          playMode: defaultPlayMode,
          theme: theme
        };

        switch (_typeof(info.musicSrc)) {
          case "function":
            info.musicSrc().then(function (val) {
              _this4.setState(_objectSpread({}, info, {
                musicSrc: val
              }, lastPlayStatus));
            }, this.onAudioLoadError);
            break;

          default:
            this.setState(_objectSpread({}, info, {}, lastPlayStatus));
        }
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._onDestroyed();

      this.unBindEvents(this.audio, undefined, false);
      this.unBindUnhandledRejection();
      this.unBindKeyDownEvents();
      this.media.removeListener(this.listenerIsMobile);
      this.media = undefined;
      this.lyric.stop();
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.addMobileListener();
      this.setDefaultAudioVolume();
      this.bindUnhandledRejection();

      if (this.props.audioLists.length >= 1) {
        this.bindEvents(this.audio);
        this.onGetAudioInstance();
        this.initLyricParser();
        this.onAddMediaSession();

        if (IS_MOBILE) {
          this.bindMobileAutoPlayEvents();
        } else {
          this.bindKeyDownEvents();
        }

        if (!IS_MOBILE && (0, _utils.isSafari)()) {
          this.bindSafariAutoPlayEvents();
        }
      }
    }
  }]);

  return ReactJkMusicPlayer;
}(_react.PureComponent);

exports["default"] = ReactJkMusicPlayer;

_defineProperty(ReactJkMusicPlayer, "defaultProps", {
  audioLists: [],
  theme: "dark",
  mode: "mini",
  defaultPlayMode: "order",
  defaultPosition: {
    left: 0,
    top: 0
  },
  once: false,
  //onAudioPlay 事件  是否只触发一次
  drag: true,
  toggleMode: true,
  //能换在迷你 和完整模式下 互相切换
  showMiniModeCover: true,
  //迷你模式下 是否显示封面图
  showDownload: true,
  showPlay: true,
  showReload: true,
  showPlayMode: true,
  showThemeSwitch: true,
  showLyric: false,
  playModeTipVisible: false,
  //手机端切换播放模式
  autoPlay: true,
  defaultVolume: 1,
  showProgressLoadBar: true,
  //音频预加载进度
  seeked: true,
  playModeShowTime: 600,
  //播放模式提示 显示时间,
  bounds: "body",
  //mini 模式拖拽的可移动边界
  showMiniProcessBar: false,
  //是否在迷你模式 显示进度条
  loadAudioErrorPlayNext: true,
  // 加载音频失败时 是否尝试播放下一首
  preload: false,
  //是否在页面加载后立即加载音频
  glassBg: false,
  //是否是毛玻璃效果
  remember: false,
  //是否记住当前播放状态
  remove: true,
  //音乐是否可以删除
  defaultPlayIndex: 0,
  //默认播放索引
  getContainer: function getContainer() {
    return document.body;
  },
  // 播放器挂载的节点
  autoHiddenCover: false,
  // 当前播放歌曲没有封面时是否自动隐藏
  onBeforeAudioDownload: function onBeforeAudioDownload() {},
  // 下载前转换音频地址等
  spaceBar: false,
  // 是否可以通过空格键 控制播放暂停
  showDestroy: false,
  showMediaSession: false,
  locale: _locale["default"].en_US
});

_defineProperty(ReactJkMusicPlayer, "propTypes", _propTypes["default"]);
