// react-jinke-music-player/lib/components/PlayerMobile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _PlayModeTip = _interopRequireDefault(require("./PlayModeTip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var prefix = 'react-jinke-music-player-mobile';

var PlayerMobile = function PlayerMobile(_ref) {
  var name = _ref.name,
      cover = _ref.cover,
      singer = _ref.singer,
      playing = _ref.playing,
      duration = _ref.duration,
      currentTime = _ref.currentTime,
      loading = _ref.loading,
      loadingIcon = _ref.loadingIcon,
      themeSwitch = _ref.themeSwitch,
      progressBar = _ref.progressBar,
      openAudioListsPanel = _ref.openAudioListsPanel,
      audioPrevPlay = _ref.audioPrevPlay,
      audioNextPlay = _ref.audioNextPlay,
      prevAudioIcon = _ref.prevAudioIcon,
      nextAudioIcon = _ref.nextAudioIcon,
      playIcon = _ref.playIcon,
      pauseIcon = _ref.pauseIcon,
      playMode = _ref.playMode,
      downloadIcon = _ref.downloadIcon,
      reloadIcon = _ref.reloadIcon,
      playListsIcon = _ref.playListsIcon,
      closeIcon = _ref.closeIcon,
      onClose = _ref.onClose,
      pause = _ref.pause,
      playModeTipVisible = _ref.playModeTipVisible,
      currentPlayModeName = _ref.currentPlayModeName,
      extendsContent = _ref.extendsContent,
      onPlay = _ref.onPlay,
      glassBg = _ref.glassBg,
      LyricIcon = _ref.LyricIcon,
      autoHiddenCover = _ref.autoHiddenCover;
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(prefix, {
      'default-bg': !glassBg,
      'glass-bg': glassBg
    })
  }, /*#__PURE__*/_react["default"].createElement(_PlayModeTip["default"], {
    prefix: prefix,
    visible: playModeTipVisible,
    title: playMode,
    text: currentPlayModeName
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-header group")
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "left item"
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: "title"
  }, name), /*#__PURE__*/_react["default"].createElement("div", {
    className: "right item",
    onClick: onClose
  }, closeIcon)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-singer text-center group")
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "name"
  }, singer)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-switch text-center group")
  }, themeSwitch), (!autoHiddenCover || autoHiddenCover && cover) && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-cover text-center")
  }, /*#__PURE__*/_react["default"].createElement("img", {
    src: cover,
    alt: "cover",
    className: (0, _classnames["default"])('cover', {
      'img-rotate-pause': pause || !playing || !cover
    })
  })), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-progress group")
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "current-time"
  }, loading ? '--' : currentTime), progressBar, /*#__PURE__*/_react["default"].createElement("span", {
    className: "duration text-right"
  }, loading ? '--' : duration)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-toggle text-center group")
  }, loading ? loadingIcon : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("span", {
    className: "group prev-audio",
    title: "Previous track",
    onClick: audioPrevPlay
  }, prevAudioIcon), /*#__PURE__*/_react["default"].createElement("span", {
    className: "group play-btn",
    title: playing ? 'Click to pause' : 'Click to play',
    onClick: onPlay
  }, playing ? pauseIcon : playIcon), /*#__PURE__*/_react["default"].createElement("span", {
    className: "group next-audio",
    title: "Next track",
    onClick: audioNextPlay
  }, nextAudioIcon))), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-operation group")
  }, /*#__PURE__*/_react["default"].createElement("ul", {
    className: "items"
  }, [playMode, downloadIcon, reloadIcon, LyricIcon].filter(Boolean).map(function (icon, i) {
    return /*#__PURE__*/_react["default"].createElement("li", {
      className: "item",
      key: i
    }, icon);
  }), extendsContent, /*#__PURE__*/_react["default"].createElement("li", {
    className: "item",
    onClick: openAudioListsPanel
  }, playListsIcon))));
};

var _default = PlayerMobile;
exports["default"] = _default;
